### PR TITLE
Shell: Implement the '$*' special variable in the POSIX parser 

### LINKED ISSUE #21421: Shell: "$*" isn't supported in Posix mode
As seen by the explicit TODO:

https://github.com/SerenityOS/serenity/blob/3e86f88d6aec078dbb533cfabcd63f13c982357a/Userland/Shell/PosixParser.cpp#L1677-1680

This breaks parsing the mksh Build.sh script here:

https://github.com/MirBSD/mksh/blob/master/Build.sh#L257-L260

Description for this option can be seen here: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_05_02

> \*
>  Expands to the positional parameters, starting from one, initially producing one field for each positional parameter that is set. When the expansion occurs in a context where field splitting will be performed, any empty fields may be discarded and each of the non-empty fields shall be further split as described in [Field Splitting](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_05). When the expansion occurs in a context where field splitting will not be performed, the initial fields shall be joined to form a single field with the value of each parameter separated by the first character of the IFS variable if IFS contains at least one character, or separated by a <space> if IFS is unset, or with no separation if IFS is set to a null string.

### PR DESCRIPTION
**Pull request by alimpfard**
Fixes #21421.
image

### CODE REVIEW

File: Userland/Shell/PosixParser.cpp [hunk 1]
```diff
@@ -1674,9 +1682,7 @@ ErrorOr<RefPtr<AST::Node>> Parser::parse_word()
                 '?');
             break;
         case ResolvedParameterExpansion::Op::GetPositionalParameterListAsString:
-            node = make_ref_counted<AST::SyntaxError>(
-                token.position.value_or(empty_position()),
-                "$* not implemented"_string);
+            node = strip_execute(::Shell::Parser { "${join \"${defined_value_or_default IFS ' '}\" $*}"sv }.parse());
```
Comment (ADKaster) [inline Userland/Shell/PosixParser.cpp:1685]: Is this supposed to respect the calling context's IFS?

Comment (ADKaster) [reply Userland/Shell/PosixParser.cpp:1685]: Guess not
```
$ set a b c
$ echo $*
a b c
$ IFS='\t' echo $*
a b c
$ export IFS='\t'
$ echo $*
a b c
$ export IFS='\n'
$ echo $*
a b c
$ 
```
Carry on then

File: Userland/Shell/Builtin.cpp [hunk 2]
```diff
@@ -332,6 +332,26 @@ ErrorOr<int> Shell::builtin_bg(Main::Arguments arguments)
...
 }
 
+ErrorOr<String> Shell::serialize_function_definition(ShellFunction const& fn) const
+{
+    StringBuilder builder;
+    builder.append(fn.name);
+    builder.append('(');
+    for (size_t i = 0; i < fn.arguments.size(); i++) {
+        builder.append(fn.arguments[i]);
+        if (!(i == fn.arguments.size() - 1))
```
[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Shell/Builtin.cpp:342]: :thinking: I'm wondering _how_ I came up with this

[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Shell/Builtin.cpp]: Oh! that code was copied from inside `builtin_type`, it was like that before.
fixed anyway.

**Pull request merged**